(Closes #1423) Blocking Loop transformation

LonelyCat124 opened the pull request:
Open the pull request the right way around doh...

Still to do:

- [x] Move transformation into its own file
- [x] Move tests
- [x] Finish testing (I expect at least the BlockedLoop node has no testing right now)
- [x] Add the transformation to the user guide

sergisiso commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+        node_sigs = node_vars.all_signatures
+
+        # Find the Loop code's signatures
+        child_refs = VariablesAccessInfo(node.children[3])
```
And maybe "bounds_ref" and "bounds_sigs" for the "node_vars" and "node_sigs" if that makes sense.

LonelyCat124 replied:
Done

sergisiso commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+            # If access2 is a write then we write to a loop variable
+            if access2.is_written():
+                raise TransformationError("Cannot apply a BlockedLoopTrans "
+                                          "to a loop where loop variables are "
```
"to this loop because the boundary variable '{0}' is written..." and then format(access2.name)

LonelyCat124 replied:
Done

sergisiso commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+        child1.replace_with(Reference(end_inner_loop))
+
+        # Create the outerloop of the same type and loop_type
+        outerloop = Loop(variable=outer_loop_variable,
```
Maybe we discussed last week and I forgot, but is there any reason this can not be `outerloop = Loop(variable=outer_loop_variable)` and then don't attempt to set the loop_type (since we already use the annotation to mark that it is a blocked loop). If that's valid, then you can remove the introduced valid_loop_types property.

LonelyCat124 replied:
Perhaps I'd misunderstood, I though we decided to create this new loop's type to be the same type as the loop that is being transformed - the idea being to enable the transformation to be able to be API agnostic (whilst still being supported).

What I wanted to do was do `outerloop = type(node).create(...)`, however the `.create` function is not at all standardised, so this was my workaround. Its not perfect, as I'm not sure how psy.gen will work for these loop types if they're supported by APIs in someway, however my expectation is that they will probably work.

sergisiso replied:
Yeah that's fine then, I just wasn't sure if we say to reuse the same loop_type before or after we said the annotations could be used to mark these loops. I am not too concerned about this as probably will be revisited after the loop standarisation to make sure everything matches.

sergisiso commented on src/psyclone/tests/psyir/transformations/block_loop_trans_test.py:
```diff
@@ -0,0 +1,360 @@
+# -----------------------------------------------------------------------------
+# Authors A. B. G. Chalk STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+'''This module contains the unit tests for the BlockLoopTrans module'''
```
Blank lines before and after the module docstring please

LonelyCat124 replied:
Done

sergisiso replied:
The *after* blank line is still missing :)

LonelyCat124 replied:
Done

sergisiso commented on src/psyclone/tests/psyir/transformations/block_loop_trans_test.py:
```diff
@@ -0,0 +1,360 @@
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE, \
+    ScalarType, SymbolTable
+from psyclone.psyir.transformations import TransformationError, BlockLoopTrans
+from psyclone.parse.algorithm import parse
+from psyclone.psyGen import PSyFactory
```
Please move this before the psyclone.psyir import for alphabetical order

LonelyCat124 replied:
Done

sergisiso commented on src/psyclone/tests/psyir/transformations/block_loop_trans_test.py:
```diff
@@ -0,0 +1,360 @@
+    blocktrans = BlockLoopTrans()
+    blocktrans.apply(schedule.children[0])
+    code = str(psy.gen)
+    print(code)
```
Delete print statement

sergisiso commented on src/psyclone/tests/psyir/transformations/block_loop_trans_test.py:
```diff
@@ -0,0 +1,360 @@
+
+
+def test_blockloop_trans_apply_pos():
+    '''Test the apply function of BlockLoopTrans'''
```
"... for a positive step." And if I have to be picky in all these file docstrings it should say method instead of function :)

LonelyCat124 replied:
Done

sergisiso commented on src/psyclone/tests/psyir/transformations/block_loop_trans_test.py:
```diff
@@ -0,0 +1,360 @@
+      enddo
+    enddo
+  enddo'''
+    assert correct in result
```
Can you add:
(at the top)
from psyclone.tests.utilities import Compile

(at the test signature)
def test_blockloop_trans_apply_double_block(tmpdir):

(and here)
assert Compile(tmpdir).string_compiles(result)

To make sure the generated Fortran is valid (you may need the "end" integer to be a parameter and other small fixes)

LonelyCat124 replied:
Done, it still just compiles when I run the test though.

sergisiso replied:
Can you bring the "assert Compile(tmpdir).string_compiles(result)" statement to the end. I think it makes sense we first check the produced code is as expected and then we try to compile.

I don't understand your response but to test it you need to do:
`pytest --compile src/psyclone/tests/psyir/transformations/block_loop_trans_test.py`
and at the moment this gives a compiler error

LonelyCat124 replied:
Sorry I missed this. It now does work with `pytest --compile`

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors A. B. G. Chalk STFC Daresbury Lab
+# -----------------------------------------------------------------------------
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+# Authors A. B. G. Chalk STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+'''This module provides the BlockLoopTrans, which transforms a Loop into a
+blocked Loop'''
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+
+class BlockLoopTrans(LoopTrans):
+    '''
+    Apply a blocking transformationt to a loop (in order to permit a
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+    Apply a blocking transformationt to a loop (in order to permit a
+    chunked parallelisation). For example:
+
+    >>> from psyclone.parse.algorithm import parse
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+
+    def validate(self, node, options=None):
+        '''
+        Validates that the Loop represented by :py:obj:`node` can have
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+        :param node: the loop to validate.
+        :type node: :py:class:`psyclone.psyir.nodes.Loop`
+        :param options: a dictionary with options for transformation.
+        :type options: dictionary of string:values or None
```
Done

sergisiso replied:
I forgot in the previous review that string should also be str in the type for now (also below in the apply method docstring)

LonelyCat124 replied:
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+                which is not a constant value.
+        :raises TransformationError: if the supplied Loop has a step size \
+                larger than the chosen block size.
+        :raises TransformationError: if the supplied Loop has is a blocked
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+                variables inside the Loop body.
+        '''
+        super(BlockLoopTrans, self).validate(node, options=options)
+        # If step is a variable don't support it for now.
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+        # If step is a variable don't support it for now.
+        if options is None:
+            options = {}
+        if isinstance(node.children[2], nodes.Reference):
```
Yeah, makes sense, done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+            raise TransformationError("Cannot apply a BlockLoopTrans to "
+                                      "a loop with a non-constant step size")
+        block_size = options.get("blocksize", 32)
+        if abs(int(node.children[2].value)) > block_size:
```
Done and tested I think now.

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+
+        if int(node.children[2].value) == 0:
+            raise TransformationError("Cannot apply a BlockLoopTrans to "
+                                      "a loop with a step size of 0")
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+        if abs(int(node.children[2].value)) > block_size:
+            raise TransformationError("Cannot apply a BlockLoopTrans to "
+                                      "a loop with larger step size than the "
+                                      "chosen block size")
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+
+    def apply(self, node, options=None):
+        '''
+        Converts the Loop represented by :py:obj:`node` into a
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+        '''
+        Converts the Loop represented by :py:obj:`node` into a
+        nested loop where the outer loop is over blocks and the inner
+        loop is over each block.
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+        :param node: the loop to transform.
+        :type node: :py:class:`psyclone.psyir.nodes.Loop`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+
+        # Store the node's parent for replacing later and the start and end
+        # indicies
+        child0 = node.children[0]
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,237 @@
+
+        # Add the access pattern to the node variable name
+        node_vars.add_access(Signature(node.variable.name),
+                             AccessType.WRITE, self)
```
I think so, doesn't appear to cause any issues.

LonelyCat124 commented:
I've merged master and updated all the comments I think

sergisiso reviewed (changes requested):
@LonelyCat124 Added some more comments and replied to some of the previous ones.

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,258 @@
+        if not isinstance(node.children[2], nodes.Literal):
+            # If step is a variable we don't support it.
+            raise TransformationError("Cannot apply a BlockLoopTrans to "
+                                      "a loop with a non-constant step size.")
```
Yes, I'll change this.

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,258 @@
+        if node.children[2].datatype.intrinsic is not \
+           ScalarType.Intrinsic.INTEGER:
+            raise TransformationError("Cannot apply a BlockLoopTrans to a "
+                                      "loop with a non-integer step size.")
```
Done

LonelyCat124 commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,258 @@
+        if int(node.children[2].value) == 0:
+            raise TransformationError("Cannot apply a BlockLoopTrans to "
+                                      "a loop with a step size of 0.")
+        # Other checks needed for validation
```
I did this for `CodeBlock` only as the others are not affected by this transformation at this point in time.

LonelyCat124 commented:
I fixed up what I could as we discussed on teams, and renamed the transformation accordingly (hopefully I didn't miss anything, tests still ran and I updated the user-guide)

sergisiso commented on src/psyclone/psyir/transformations/block_loop_trans.py:
```diff
@@ -0,0 +1,258 @@
+    '''
+    Apply a blocking transformation to a loop (in order to permit a
+    chunked parallelisation or improve cache utilisation). For example:
+
```
Aha, its that the ".. code-block:: fortran" must be preceded by a blank line otherwise it doesn't parse the text inside as a "code-block". When adding the bank line all warnings disappear.

LonelyCat124 replied:
Ok, I've fixed that now too

sergisiso commented on src/psyclone/tests/psyir/transformations/chunk_loop_trans_test.py:
```diff
@@ -0,0 +1,401 @@
+# -----------------------------------------------------------------------------
```
Same in this file:
```
$ grep -n "block" src/psyclone/tests/psyir/transformations/chunk_loop_trans_test.py
58:def test_blockloop_trans():
61:    assert str(trans) == "Split a loop into a blocked loop pair"
64:def test_blockloop_trans_validate1():
90:def test_blockloop_trans_validate2():
100:        chunktrans.validate(parent, {"blocksize": 16})
102:            "(128) than the chosen block size (16).") in str(excinfo.value)
116:def test_blockloop_trans_validate3():
118:    to a loop which has already been blocked'''
121:    # parent loop (can't apply a block loop trans to a block loop trans
137:#    assert "Cannot apply a ChunkLoopTrans to an already blocked loop." \
141:    assert "Cannot apply a ChunkLoopTrans to an already blocked loop." \
145:def test_blockloop_trans_validate4():
232:def test_blockloop_trans_validate5():
261:def test_blockloop_trans_validate6():
278:def test_blockloop_trans_validate7():
296:def test_blockloop_trans_apply_pos():
319:def test_blockloop_trans_apply_neg():
343:def test_blockloop_trans_apply_double_block(tmpdir):
345:    blocks of 2 nested loops'''
```

LonelyCat124 replied:
Think I got them all now

sergisiso commented on src/psyclone/tests/psyir/transformations/chunk_loop_trans_test.py:
```diff
@@ -0,0 +1,401 @@
+                    ScalarType.Precision.SINGLE))
+    parent._variable = lvar
+    chunktrans.apply(parent)
+#    with pytest.raises(TransformationError) as excinfo:
+#        chunktrans.validate(parent.ancestor(Loop))
+#    assert "Cannot apply a ChunkLoopTrans to an already blocked loop." \
+#           in str(excinfo.value)
```
Should this be uncommented or removed?

LonelyCat124 replied:
It should be uncommented, I'd commented it out before when i was fixing tests but it seems fine now.

sergisiso commented on src/psyclone/tests/psyir/transformations/chunk_loop_trans_test.py:
```diff
@@ -0,0 +1,401 @@
+    correct = '''END DO
+        END DO
+      END DO'''
+    assert correct in code
```
Can you explicitly assert the the "blocked" (I think now "chunked") annotation has been added to the loop.

LonelyCat124 replied:
Added

sergisiso reviewed (changes requested):
@LonelyCat124 Thanks for renaming the transformation, it makes it more clear what it does. There are still a few "block/blocking" mentions in the code that need updating.

Also, check the previous review comments as there are still unaddressed comments. (You may need to click the "load more" if they are not displayed). Specifically the one regarding:
 `pytest --compile src/psyclone/tests/psyir/transformations/chunk_loop_trans_test.py`
needs to be solved as passing tests with compilation is part of the review process.

LonelyCat124 commented on src/psyclone/psyir/transformations/chunk_loop_trans.py:
```diff
@@ -0,0 +1,272 @@
+
+class ChunkLoopTrans(LoopTrans):
+    '''
+    Apply a blocking transformation to a loop (in order to permit a
```
Think I got them all now

sergisiso commented:
See the second part of my review comment, you still need to address the open issues from previous reviews  :)

LonelyCat124 commented:
Sorry, I didn't see the previous comments my bad. I think they're all fixed up now.

sergisiso reviewed (approved):
@LonelyCat124 All the comments have been addressed. Tests and examples pass with compilation. There are no relevant pycodestyle or pylint issues introduced in this PR. To TODOs referencing this issue. And user/developer_guide built without new warnings.

This is ready to be merged.
